programmer-skills/http-intro.md

```diff
@@ -0,0 +1,361 @@
+
+#####302 Found
+
+A temporary redirection. Future requests may continue using the original URL, not the temporary one returned in the request. As with `302`, browsers are liable to change the original request method on redirect.
+
+#####303 See Other
+
+This is primarily intended for responding to `POST` requests. Remember that a server may respond to a `POST` request by storing the resource wherever it wants. It provides a `303 See Other` response to let the client know the resource's URL so that the user may bookmark it etc. 
+
+#####304 Not Modified
+
+This is used along with caching headers to indicate that the resource has not been modified and the client's cached version should be used.
+
+#####307 Temporary Redirect
+
+This status code is similar to `302` but corrects the error that browsers may switch HTTP methods. A client browser may not switch method to follow a `307` redirect.
+
+#####308 Permanent Redirect
+
+As with `307`, this is similar to `301` but browsers may not switch methods.
+
+####4xx
+
+These codes indicate some kind of error with the request.
+
+#####400 Bad Request
+
+The request was malformed in some way and the server could not understand it.
+
+#####401 Unauthorized
+
+The request needs to be authenticated. The response will include a `WWW-Authenticate` header specifying how to authenticate.
+
+#####403 Forbidden
+
+The server understood what you requested but refuses to authorise it. It may give the reason in the response payload but it is not required to.
+
+#####404 Not Found
+
+The famous one! The client has requested a resource that the server doesn't know how to find.
+
+#####418 I'm a Teapot
+
+Any attempt to brew coffee with a teapot should result in the error code `418 I'm a teapot`. The resulting entity body MAY be short and stout.
+
+#####429 Too Many Requests
+
+The server is rate-limiting the client's requests and too many have been sent. The server may optionally provide a `Retry-After` headers saying how long to wait.
+
+####5xx
+
+This group of responses indicates an error with the server.
+
+#####500 Internal Server Error
+
+The server has encountered a problem and can't fulfil the request. Normally this happens if the route handler has a bug in it.
+
+#####503 Service Unavailable
+
+The server can't deal with your request due to temporary overload or scheduled maintenance. It might respond with a `Retry-After` header to let the client know how long to wait.
+
+Related
+=======
+
+Now that you have a better understanding of how HTTP works, you can apply it to [setting up a local server](http://codingforeveryone.foundersandcoders.org/programmer-skills/setting-up-a-local-server.html).
+
+Further reading
+===============
+
+https://httpstatuses.com/
+
+http://code.tutsplus.com/tutorials/http-headers-for-dummies--net-8039
+
+[The actual specification](https://www.w3.org/Protocols/rfc2616/rfc2616.txt), if you're feeling brave!
+
+TODO
```

Comment:
I have seen other sections that have a TODO section at the bottom and I think it's a useful indicator for anyone looking for something to contribute, so I was planning to keep it in.